File: daily_python/2044_Count_Number_of_Maximum_Bitwise_OR_Subsets.py
import os
import sys
from typing import List

class Solution:
    def dfs(self, nums, i, curr, target):
        if i == len(nums):
            return
        self.dfs(nums, i+1, curr, target)
        curr = curr | nums[i]
        if (curr == target):
            self.ans += (1 << (len(nums)-1-i))
        else:
            self.dfs(nums, i+1, curr, target)

    def countMaxOrSubsets(self, nums: List[int]) -> int:
        self.ans = 0
        target = 0
        for i in nums:
            target = target | i
        self.dfs(nums, 0, 0, target)
        return self.ans
    
def readDataSet(filename):
    dataset = []
    with open(filename, 'r') as file:
        content = file.read().strip()
        blocks = content.split('\n\n')
        for block in blocks:
            lines = block.split('\n')
            nums = list(map(int, lines[0].split('=')[1].strip()[1:-2].split(',')))
            dataset.append(nums)
    return dataset

if __name__ == '__main__':
    if len(sys.argv) == 1:
        filename = os.path.splitext(__file__)[0] +'.txt'
    else:
        filename = sys.argv[1]
    dataset = readDataSet(filename)
    results = []
    solution = Solution()
    for nums in dataset:
        results.append(solution.countMaxOrSubsets(nums))
    for index, result in enumerate(results):
        print(f'Example {index + 1} : {result}')

File: daily/2024/2044_Count_Number_of_Maximum_Bitwise_OR_Subsets.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
    int ans = 0;
    void dfs(vector<int>& nums, int i, int curr, int target){
        if (i == nums.size()) return;
        dfs(nums, i+1, curr, target);
        curr |= nums[i];
        if (curr == target) ans += 1 << (nums.size() - 1 - i);
        else{
            dfs(nums, i+1, curr, target);
        }
    }
    int countMaxOrSubsets(vector<int>& nums) {
        int n = nums.size();
        int target = 0;
        for (int i : nums) target |= i;
        dfs(nums, 0, 0, target);
        return ans;
    }
};

int main(){
    vector<int> nums = {3,1};
    // vector<int> nums = {2,2,2};
    // vector<int> nums = {3,2,1,5};
    Solution S;
    cout << S.countMaxOrSubsets(nums) << endl;
    return 0;
}